redskyapi/experiments/v1alpha1/trial.go

```diff
@@ -0,0 +1,146 @@
+	"net/url"
+	"strings"
+	"time"
+)
+
+type TrialMeta struct {
+	SelfURL   string `json:"-"`
+	LabelsURL string `json:"-"`
+}
+
+func (m *TrialMeta) SetLocation(location string) { m.SelfURL = location }
+func (m *TrialMeta) SetLastModified(time.Time)   {}
+func (m *TrialMeta) SetLink(rel, link string) {
+	switch rel {
+	case relationLabels:
+		m.LabelsURL = link
+	}
+
+	// Backwards compatibility with the old trial labels relation
+	if m.LabelsURL == "" && rel == "https://carbonrelay.com/rel/trialLabels" {
+		m.LabelsURL = link
+	}
+}
+
+type Assignment struct {
+	// The name of the parameter in the experiment the assignment corresponds to.
+	ParameterName string `json:"parameterName"`
+	// The assigned value of the parameter.
+	Value json.Number `json:"value"`
+}
+
+type TrialAssignments struct {
+	TrialMeta
+
+	// The list of parameter names and their assigned values.
+	Assignments []Assignment `json:"assignments"`
+}
+
+type Value struct {
+	// The name of the metric in the experiment the value corresponds to.
+	MetricName string `json:"metricName"`
+	// The observed value of the metric.
+	Value float64 `json:"value"`
+	//The observed error of the metric.
+	Error float64 `json:"error,omitempty"`
+}
+
+type TrialValues struct {
+	// The observed values.
+	Values []Value `json:"values,omitempty"`
+	// Indicator that the trial failed, Values is ignored when true.
+	Failed bool `json:"failed,omitempty"`
+}
+
+type TrialStatus string
+
+const (
+	TrialStaged    TrialStatus = "staged"
+	TrialActive    TrialStatus = "active"
+	TrialCompleted TrialStatus = "completed"
+	TrialFailed    TrialStatus = "failed"
+	TrialAbandoned TrialStatus = "abandoned"
+)
+
+type TrialItem struct {
+	TrialAssignments
+	TrialValues
+
+	// The current trial status.
+	Status TrialStatus `json:"status"`
+	// Ordinal number indicating when during an experiment the trail was generated.
+	Number int64 `json:"number"`
+	// Labels for this trial.
+	Labels map[string]string `json:"labels,omitempty"`
+
+	// The metadata for an individual trial.
+	Metadata Metadata `json:"_metadata,omitempty"`
+
+	// Experiment is a reference back to the experiment this trial item is associated with. This field is never
+	// populated by the API, but may be useful for consumers to maintain a connection between resources.
+	Experiment *Experiment `json:"-"`
+}
+
+type TrialListQuery struct {
+	// Comma separated list of statuses to fetch.
+	Status []TrialStatus
+	// Comma separated list of label value pairs to match on.
+	LabelSelector map[string]string
+}
+
+func (p *TrialListQuery) Encode() string {
+	if p == nil {
+		return ""
+	}
+	q := url.Values{}
+	if len(p.Status) > 0 {
+		strs := make([]string, len(p.Status))
+		for i := range p.Status {
+			strs[i] = string(p.Status[i])
+		}
+		q.Add("status", strings.Join(strs, ","))
+	}
+	if len(p.LabelSelector) > 0 {
+		ls := make([]string, 0, len(p.LabelSelector))
+		for k, v := range p.LabelSelector {
+			ls = append(ls, fmt.Sprintf("%s=%s", k, v))
```

Comment:
not sure how much of this is carry over, but should we use https://golang.org/pkg/net/url/#QueryEscape instead of `fmt.Sprintf`?

Reply:
That would double encode, i.e. this function would return `labelSelector=foo%253Dbar` instead of just `labelSelector=foo%3Dbar`.

This output gets dropped into `URL.RawQuery` which does the right thing (which isn't really the right thing because technically `=` doesn't need to be encoded here but it's better to error on the side of caution because some URL parsers are just garbage).